src/view/treeNodes/fileChangeNode.ts

```diff
@@ -211,6 +211,117 @@ export class GitFileChangeNode extends TreeNode implements vscode.TreeItem {
 	}
 }
 
+export class DirectoryTreeNode extends TreeNode implements vscode.TreeItem {
+	public collapsibleState: vscode.TreeItemCollapsibleState;
+	public children: (RemoteFileChangeNode | InMemFileChangeNode | GitFileChangeNode| DirectoryTreeNode)[] =  new Array();
+	private pathToChild: Map<string, DirectoryTreeNode> = new Map();
+
+	constructor(
+		public parent: TreeNode | vscode.TreeView<TreeNode>,
+		public label: string,
+	) {
+		super();
+		this.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+	}
+
+	async getChildren(): Promise<TreeNode[]> {
+		return this.children;
+	}
+
+	public trimTree(): void {
+		if (this.children.length === 0) {
+			return;
+		}
+
+		this.children.forEach(n => {
+			if (n instanceof DirectoryTreeNode) {
+				n.trimTree(); // reccursive
+			}
+		});
+
+		// merge if this only have single directory, eg:
+		// - a
+		//   - b
+		//     - c
+		// becomes:
+		// - a/b
+		//   - c
+		if (this.children.length !== 1) {
+			return;
+		}
+		let child = this.children[0];
+		if (!(child instanceof DirectoryTreeNode)) {
+			return;
+		}
+
+		// perform the merge
+		this.label = this.label + '/' + child.label;
+		if (this.label.startsWith('/')) {
+			this.label = this.label.substr(1);
+		}
+		this.children = child.children;
+	}
+
+	sort(): void {
+		if (this.children.length <= 1) {
+			return;
+		}
+
+		let dirs: DirectoryTreeNode[] = [];
+		let files: (RemoteFileChangeNode | InMemFileChangeNode | GitFileChangeNode)[] = [];
+
+		// process directory
+		this.children.forEach(node => {
+			if (node instanceof DirectoryTreeNode) {
+				node.sort(); // recc
+				dirs.push(node);
+			} else {
+				// files
+				files.push(node);
+			}
+		});
+
+		// sort
+		dirs.sort((a, b) => (a.label < b.label) ? -1 : 1);
+		files.sort((a, b) => (a.label < b.label) ? -1 : 1);
+
+		this.children = [...dirs, ...files];
+	}
+
+	public addFile(file: RemoteFileChangeNode | InMemFileChangeNode): void {
+		let paths = file.fileName.split('/');
```

Comment:
nit: we're not consistent about this style everywhere in the repo, but I would prefer using `const` over `let` when possible